### PR TITLE
Implement tx counter for transaction info

### LINKED ISSUE #601: Add transaction index implemented as counter
We'd like to expose the index of a transaction in a block in https://github.com/CosmWasm/cosmwasm/pull/1077. This allows contracts to get a pair `(block height, transaction index)` that uniquely identifies the transaction on a given chain. This transaction index should be easy to work with in other tools, e.g. by looking in a block explorer or using a call like

```
curl -sS "https://rpc.cosmos-hub.app.beta.starport.cloud/block?height=7590762" | jq .result.block.data.txs
```

In https://github.com/CosmWasm/wasmd/blob/v0.18.0/proto/cosmwasm/wasm/v1/types.proto#L123-L131 the gas meter is used for ordering, but the values from this are hard to work with. So what we look for is an 0-based index 0, 1, 2, ….

### PR DESCRIPTION
**Pull request by alpe**
Resolves #601 

Based on the #608 spike

Changes: 
- Use `uint32` for the tx counter
- Sets wasmvm TransactionInfo
- Non default `0` counter but empty TransactionInfo on simulations
- Tests
- Better doc

### CODE REVIEW

File: app/ante.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+		ante.NewValidateMemoDecorator(ak),
+		ante.NewConsumeGasForTxSizeDecorator(ak),
+		ante.NewRejectFeeGranterDecorator(),
+		ante.NewSetPubKeyDecorator(ak), // SetPubKeyDecorator must be called before all signature verification decorators
+		ante.NewValidateSigCountDecorator(ak),
+		ante.NewDeductFeeDecorator(ak, bankKeeper),
+		ante.NewSigGasConsumeDecorator(ak, sigGasConsumer),
+		ante.NewSigVerificationDecorator(ak, signModeHandler),
+		ante.NewIncrementSequenceDecorator(ak),
+		wasmkeeper.NewCountTXDecorator(txCounterStoreKey),
```
Comment (ethanfrey) [inline app/ante.go:36]: We should add some note that chains using x/wasm need to add this custom AnteHandler if they want Tx Index info.

Not sure where that goes? Integration.md?

**Review by ethanfrey — APPROVED** [review]
Nice clean version.

:+1:

**Comment by ethanfrey** [comment]
Please make sure to tick the box in https://github.com/CosmWasm/wasmd/issues/618 after merging

File: INTEGRATION.md [hunk 2]
```diff
@@ -63,6 +63,10 @@ from the Cosmos SDK, and enabled them in `app.go`. If so, you can just look
 at [`wasmd/app/app.go`](https://github.com/CosmWasm/wasmd/blob/master/app/app.go#)
 for how to do so (just search there for lines with `wasm`).
 
+`wasmd` also comes with a custom `ante handler` that adds the TX position in the block into the context
```
Comment (ethanfrey) [inline INTEGRATION.md:66]: Nice

File: contrib/local/02-contracts.sh [hunk 3]
```diff
@@ -8,7 +8,7 @@ echo "## Add new CosmWasm contract"
 RESP=$(wasmd tx wasm store "$DIR/../../x/wasm/keeper/testdata/hackatom.wasm" \
   --from validator --gas 1500000 -y --chain-id=testing --node=http://localhost:26657 -b block)
 
-CODE_ID=$(echo "$RESP" | jq -r '.logs[0].events[0].attributes[-1].value')
+CODE_ID=$(echo "$RESP" | jq -r '.logs[0].events[1].attributes[-1].value')
```
Comment (ethanfrey) [inline contrib/local/02-contracts.sh:11]: The ante handler added an event?
Or this is just an older bug you just found?

**Review by ethanfrey — APPROVED** [review]
lgtm

File: contrib/local/02-contracts.sh (hunk 3 above)
Comment (alpe) [reply contrib/local/02-contracts.sh:11]: This was a 🐛 that I found with manual tests. Not related to the ante handler.

**Pull request merged**